### ISSUE TITLE
Retrieving informational messages returned by server

### ISSUE THREAD

**Issue by gordthompson** [mkleehammer/pyodbc#495]
(Inspired by [this Stack Overflow question](https://stackoverflow.com/q/53778988/2144390).)

Is there a way that we can retrieve informational messages sent by the server, for example, those produced by a `PRINT` statement in a T-SQL stored procedure?

For a test SP

```sql
USE [myDb]
GO

/****** Object:  StoredProcedure [dbo].[PrintTest]    Script Date: 2018-12-14 10:59:12 ******/
SET ANSI_NULLS ON
GO

SET QUOTED_IDENTIFIER ON
GO

-- =============================================
-- Author:		gord
-- Create date: 
-- Description:	
-- =============================================
ALTER PROCEDURE [dbo].[PrintTest] 
AS
BEGIN
    SET NOCOUNT ON;
    PRINT 'no es bueno';
END

GO
```

the message does appear in the ODBC trace

```
main            1c0c-18b0	ENTER SQLExecDirectW 
        HSTMT               0x000000000040DFA0
        WCHAR *             0x0000000002F52830 [      18] "EXEC dbo.PrintTest"
        SDWORD                    18

main            1c0c-18b0	EXIT  SQLExecDirectW  with return code 1 (SQL_SUCCESS_WITH_INFO)
        HSTMT               0x000000000040DFA0
        WCHAR *             0x0000000002F52830 [      18] "EXEC dbo.PrintTest"
        SDWORD                    18

        DIAG [01000] [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]no es bueno (0) 
```

but is there a way we can retrieve it in our Python script?

**Comment by mkleehammer** [comment]
I don't see why not - isn't that just another diagnostics record just like we get for exceptions?  I could see having a method to return an array of diagnostic messages.

What format would be best for returning them.  In the spec, they have SQLSTATE codes, text, and native error codes.

If I was using the API, I'd just want the text if I knew it was a print statement, but nothing here really tells us that.  SQLSTATE 01000 is a warning with no further classification IIRC.  I suppose either a new object or a tuple.  An object would allow us to extend it in the future.  (It needs to be something available in Python 2.7 until we finally stop supporting that.  It's EOL is coming up soon.)

**Comment by gordthompson** [comment]
My suggestion would be to return the diagnostic records in the most complete and granular form, perhaps a list of tuples something like

```
[(Msg, SqlState, NativeError), ...]
```

The `SqlState` and `NativeError` values may not be useful for retrieving T-SQL `PRINT` output but they may have some utility for handling other situations and/or ODBC drivers.

Also, re: Python_2.x support, I've been tempted to open a separate discussion about that as I've worked with the tests2/ and tests3/ versions of sqlservertests.py. My personal preference would be to concentrate solely on Python_3 since that's all I've used for the past while. However, I understand how long legacy systems stick around (even, it seems, among "rocket scientists" :) ).

**Comment by matthewgdv** [comment]
Is this still being looked at? This would be dead useful for me, as my organization uses stored procs all over the place that feed back information about the success/failure state of the subprocedures they call as print statements rather than raising errors, since an error would kill the whole thing. I currently have no way of finding out what a proc actually did when I call it using pyodbc.

**Comment by sivaprakashmp** [comment]
I have same problem. Some of stored procedures that I use returns Success, Failure & Warning messages as print output messages. They don't every raise any error or set @@error parameter. Definitely need to capture and check the messages to decide if the procedure execution was success / failure. 

I saw previous discussion about adding a new tuple that will return the messages. Yes, having a tuple with all granular details from database would be good solution & i believe this tuple will be then attached to cursor.messages / connection.messages based on  execute / connect method?

I would be happy to see if some actions taken for this enhancement request.

Otherwise, I am thinking of using native isql command as an alternate approach to read print messages from TSQL command. Any other suggestions here?

Thanks to all for looking into this.

**Comment by noisten-js** [comment]
Adding to the sentiment of how useful and appreciated this feature would be... I'm brand new to Python so I know very little. The procedures I write for applications to call include useful print statements for logging... something like this:

2020-04-07 17:53:03.5493090 | INFO | Execute [dbo].[spTest] Initializing stuff
2020-04-07 17:53:03.5493090 | INFO | Execute [dbo].[spTest] Validating parameters and stuff
2020-04-07 17:53:03.5493090 | INFO | Execute [dbo].[spTest] Doing stuff that will change data
2020-04-07 17:53:03.6672522 | INFO | Execute [dbo].[spTest] Doing stuff that will change data [ 109 ]

A message before a row affecting **set based** operation and a message after with the number of affected rows as well as other logging info. It would be great to be able to return these messages to my Python apps and write these to the app logs. As it stands only our .NET apps or SQL Agent Jobs can include these in the logs...

**Comment by jerrycl** [comment]
I agree this would be a great feature.  

PEP-249 specifies an Optional DB API Extension, of Cursor.messages, which would contain a list of tuples (exception class, exception value). This is a bit different from the list of tuples [(SqlState,Msg,NativeError)] proposed above, but the SqlState and the NativeError could be put into the exception class.

**Comment by ValerioParisMitritsakis** [comment]
Just came in to say that this would be a useful feature.

**Comment by gordthompson** [comment]
Fixed in #765
